tests/integration/test_kfp_functional_v1.py

```diff
@@ -80,28 +79,37 @@ async def test_build_and_deploy(ops_test: OpsTest, request, lightkube_client):
     bundlefile_path = Path(request.config.getoption("bundle"))
     basedir = Path("./").absolute()
 
-    # Build the charms we need to build only if --no-build is not set
     context = {}
-    if not no_build:
+
+    # Find charms in the expected path if `--charms-path` is passed
+    if charms_path := request.config.getoption("--charms-path"):
+        for charm in KFP_CHARMS:
+            cached_charm = f"{charms_path}/{charm}/{charm}_ubuntu@20.04-amd64.charm"
+            charm_resources = get_resources_from_charm_file(cached_charm)
+            context.update([(f"{charm.replace('-', '_')}_resources", charm_resources)])
+            context.update([(f"{charm.replace('-', '_')}", cached_charm)])
+    # Otherwise build the charms with ops_test
+    else:
         charms_to_build = {
             charm: Path(CHARM_PATH_TEMPLATE.format(basedir=str(basedir), charm=charm))
             for charm in KFP_CHARMS
         }
         log.info(f"Building charms for: {charms_to_build}")
-        built_charms = await ops_test.build_charms(*charms_to_build.values())
-        log.info(f"Built charms: {built_charms}")
-
-        for charm, charm_file in built_charms.items():
-            charm_resources = get_resources_from_charm_file(charm_file)
-            context.update([(f"{charm.replace('-', '_')}_resources", charm_resources)])
-            context.update([(f"{charm.replace('-', '_')}", charm_file)])
+        
+        # Build charms sequentially
+        for charm_name, charm_path in charms_to_build.items():
+            log.info(f" Building charm {charm_name}")
+            built_charm = await ops_test.build_charm(charm_path)
+            charm_resources = get_resources_from_charm_file(built_charm)
+            context.update([(f"{charm_name.replace('-', '_')}_resources", charm_resources)])
+            context.update([(f"{charm_name.replace('-', '_')}", built_charm)])
```

Comment:
nit: this block of code is a candidate for a tiny helper method. this would avoid some of the duplication with the operation in L88

Reply:
updated